c7n/resources/s3.py

```diff
@@ -707,10 +711,49 @@ def process(self, buckets):
                 results.append(f.result())
             return filter(None, results)
 
-    def process_bucket(self, func, bucket, account_id, session_factory):
-        from c7n.mu import BucketNotification
-        source = BucketNotification(
-            {'account_s3': account_id}, session_factory, bucket)
+    def process_bucket(self, func, bucket, via, account_id, session_factory):
+        if via == 'bucket-notification':
+            from c7n.mu import BucketNotification
+            source = BucketNotification(
+                {'account_s3': account_id}, session_factory, bucket)
+        elif via == 'sns':
+            from c7n.mu import SNSSubscription
+            session = session_factory()
+            s3 = session.client('s3')
+
+            # Find or create an SNS topic.
+            topic_arn = None
+            bname = bucket['Name']
+            notifications = s3.get_bucket_notification_configuration(
+                Bucket=bname)
+            all_topics = notifications.get('TopicConfigurations', [])
+            usable_topics = [t for t in all_topics
+                             if 's3:ObjectCreated:*' in t['Events']]
+            if usable_topics:
+                # Just attach to the first one, I guess?
+                topic_arn = usable_topics[0]['TopicArn']
+            else:
+                sns = session.client('sns')
+                topic_arn = sns.create_topic(Name=bname)['TopicArn']
+                policy = {
+                    'Statement': [{
+                        'Action': 'SNS:Publish',
+                        'Effect': 'Allow',
+                        'Resource': topic_arn,
+                        'Principal': {'Service': 's3.amazonaws.com'}}]}
+                sns.set_topic_attributes(
+                    TopicArn=topic_arn,
+                    AttributeName='Policy',
+                    AttributeValue=json.dumps(policy))
+                session.resource('s3').BucketNotification(bname).put(
+                    NotificationConfiguration={'TopicConfigurations': [{
```

Comment:
just realized this is an issue as well with the extant lambda function notification setup, in that we're effectively blanking extant notify config :-(